community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -54,21 +54,91 @@
 TOT_REQ_CNT = 1000
 _MAINTENANCE_SBATCH_SCRIPT_PATH = dirs.custom_scripts / "perform_maintenance.sh"
 
-NodeStatus = Enum(
-    "NodeStatus",
-    (
-        "orphan",
-        "power_down",
-        "preempted",
-        "restore",
-        "resume",
-        "terminated",
-        "unbacked",
-        "unchanged",
-        "unknown",
-    ),
-)
+class NodeAction(Protocol):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        ...
+
+    def __hash__(self):
+        return hash(type(self))
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+class NodeActionResume(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """resume nodes via scontrol"""
+        log.info(f"{count} instances to resume ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_up")
+
+class NodeActionIdle(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """idle nodes"""
+        log.info(f"{count} nodes to idle ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=resume")
+
+class NodeActionRestart(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """start instances for nodes"""
+        log.info(f"{count} instances restarted ({hostlist})")
+        start_instances(nodes)
+
+class NodeActionDown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int, reason:str="Instance stopped/deleted") -> None:
+        """down nodes"""
+        log.info(
+            f"{count} nodes set down due to node status '{status.name}' ({hostlist})"
+        )
+        run(
+            f"{lookup().scontrol} update nodename={hostlist} state=down reason='{reason}'"
+        )
+
+class NodeActionPowerDown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """power_down node in slurm"""
+        log.info(f"{count} instances to power down ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_down")
+
+class NodeActionDelete(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """delete instances for nodes"""
+        log.info(f"{count} instances to delete ({hostlist})")
+        delete_instances(nodes)
+
+class NodeActionUnknown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """Error status, nodes shouldn't get in this status"""
+        log.error(f"{count} nodes have unexpected status: ({hostlist})")
+        first = next(iter(nodes))
+        state = lookup().slurm_node(first)
+        state = "{}+{}".format(state.base, "+".join(state.flags)) if state else "None"
```

Comment:
Instead of using "model", it would be nice to make "state" a parameter of action.

Reply:
wouldn't we have to make `nodes` a parameter to achieve this? Since state depends on the first node which depends on the list of nodes passed into it.

Reply:
Discussed offline added state as parameters